src/delegate/utils.js

```diff
@@ -0,0 +1,187 @@
+const bn = require('bignumber.js')
+const tokenMetadata = require('../tokens')
+const _ = require('lodash')
+
+bn.set({ DECIMAL_PLACES: 18, ROUNDING_MODE: 4 })
+
+// TODO: work through & apply this optimizing function after fast-check has been integrated in the tests (https://github.com/dubzzz/fast-check)
+// const PRECISION = 128
+// const precision = str => {
+//   const precisionStr = bn(str).toFixed(PRECISION)
+//   const cutInsignificantDecimals = precisionStr.slice(0, precisionStr.length - 32)
+//   const trimZero = _.trimEnd(cutInsignificantDecimals, '0') || '0'
+//   const trimDot = _.trimEnd(trimZero, '.')
+//   return trimDot
+// }
+
+const precision = str => (str.toString ? str.toString() : str)
+
+/** ***
+ * MAPPING DISPLAY PRICES TO CONTRACT PRICES
+ */
+
+function getDisplayAmountsFromDisplayPrice({ senderToken, signerToken, senderAmountDisplayValue, priceDisplayValue }) {
+  let signerAmountDisplayValue
+
+  if (tokenMetadata.isBaseAsset(senderToken, [senderToken, signerToken])) {
+    signerAmountDisplayValue = precision(bn(senderAmountDisplayValue).div(priceDisplayValue))
+  } else if (tokenMetadata.isBaseAsset(signerToken, [senderToken, signerToken])) {
+    signerAmountDisplayValue = precision(bn(senderAmountDisplayValue).mul(priceDisplayValue))
+  } else {
+    throw new Error('unable to calculate baseAsset')
+  }
+
+  return { senderToken, signerToken, senderAmountDisplayValue, signerAmountDisplayValue }
+}
+
+function getAtomicAmountsFromDisplayAmounts({
+  senderToken,
+  signerToken,
+  senderAmountDisplayValue,
+  signerAmountDisplayValue,
+}) {
+  const senderAmountAtomic = tokenMetadata.formatAtomicValueByToken({ address: senderToken }, senderAmountDisplayValue)
+  const signerAmountAtomic = tokenMetadata.formatAtomicValueByToken(
+    //eslint-disable-line
+    { address: signerToken },
+    signerAmountDisplayValue,
+  )
+  return { senderToken, signerToken, senderAmountAtomic, signerAmountAtomic }
+}
+
+function getAtomicPriceFromAtomicAmounts({ senderToken, signerToken, senderAmountAtomic, signerAmountAtomic }) {
+  const priceAtomic = bn(senderAmountAtomic).div(signerAmountAtomic)
+  return { senderToken, signerToken, senderAmountAtomic, atomicPrice: priceAtomic.toString() }
+}
+
+function getContractPriceFromAtomicPrice({ senderToken, signerToken, senderAmountAtomic, atomicPrice }) {
+  const [int, decimalVal] = atomicPrice.split('.') // eslint-disable-line'
+  const decimal = decimalVal || ''
+  let priceExp
+  if (decimal === '') {
+    priceExp = 0
+  } else {
+    priceExp = decimal.length
+  }
+  const priceCoef = _.trimStart(`${int}${decimal}`, '0')
+  // check
+  const priceCheck = bn(priceCoef)
```

Comment:
If I'm understanding correctly, this is turning a number like `1234.5532` to `12345532` and multiplying by `10^4`?

Reply:
@syjk129 , this just reverses the code to check against the input, so it takes `12345532 * 10^(-4)` => `1234.5532`

Reply:
Ahh I didn't see the negative exponent. Thanks!